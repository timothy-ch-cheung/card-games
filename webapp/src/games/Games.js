import React, {useEffect, useState} from "react";
import LobbyCard from "../components/lobby-card/LobbyCard";
import styled from "styled-components";
import CreateGame from "../components/create-game/CreateGame";
import Button from "react-bootstrap/Button";
import API from "../API";
import CreatePlayer from "../components/create-player/CreatePlayer";
import {useDispatch} from "react-redux";
import {setGame} from "../actions";
import {useHistory} from "react-router-dom";
import '../App.css';
import Spinner from "react-bootstrap/Spinner";

const Divider = styled.div`
border-bottom: 1px solid grey;
margin: 10px;
`;

function Games() {
    const dispatch = useDispatch();
    const history = useHistory();
    const [showCreateGameModal, setShowCreateGameModal] = useState(false);
    const [showCreatePlayerModal, setShowCreatePlayerModal] = useState(false);
    const [showRefreshSpinner, setShowRefreshSpinner] = useState(false);
    const [currentGameId, setCurrentGameId] = useState(null);
    const [games, setGames] = useState([]);

    useEffect(() => {
        const getGames = () => {
            API.get("/games").then(function (response) {
                setGames(response.data.games)
            })
        };
        getGames();
        const interval = setInterval(function () {
            getGames();
        }, 10000)

        return () => {
            clearInterval(interval)
        }
    }, []);

    const joinGame = (gameId, userId) => {
        API.patch(`/join/${gameId}`, {
            id: userId
        }).then(function (response) {
            dispatch(setGame(gameId));
            history.push('/current-game')
        }).catch(function (error) {
            console.log(error);
        });
    }

    const handleShowCreateGameModal = () => {
        setShowCreateGameModal(true);
    }

    const handleHideCreateGameModal = () => {
        setShowCreateGameModal(false);
    }

    const handleShowCreatePlayerModal = (gameId) => {
        setCurrentGameId(gameId);
        setShowCreatePlayerModal(true);
    }

    const handleHideCreatePlayerModal = () => {
        setShowCreatePlayerModal(false);
    }

    const handleClickRefresh = () => {
        let startTime = new Date();
        setShowRefreshSpinner(true);
        API.get("/games").then(function (response) {
            setGames(response.data.games);
        }).then(function () {
            let currentTime = new Date();
            let timeElapsed = currentTime - startTime;
            if(timeElapsed > 500) {
                setShowRefreshSpinner(false);
            }
            else {
                setTimeout(function() {setShowRefreshSpinner(false);}, 500 - timeElapsed);
            }
        })
    }

    const renderCard = (card, index) => {
        return <LobbyCard gameId={card.id} lobbyName={card.lobbyName} host={card.host.username} key={index}
                          onSubmit={joinGame} showModal={handleShowCreatePlayerModal}/>
    }

    return (
        <div>
            <style type="text/css">
                {`
                        .left-title {
                            text-align: left;
                        }
                        
                        .btn-xl {
                            font-size: 1.5rem;
                            border: grey solid 1px;
                            width: 200px;
                            margin: 20px 10px 30px;
                        }
                        
                        .container {
                            max-width: 98%;
                            margin: 0 auto;
                        }
                        
                        .row {
                            display: flex;
                            flex-flow: row wrap;
                        }

                    `}
            </style>
            <Button size="xl" variant="info" onClick={e => {
                handleShowCreateGameModal();
            }}>Create Game</Button>
            <CreateGame show={showCreateGameModal} onClose={handleHideCreateGameModal}/>
            <CreatePlayer show={showCreatePlayerModal} onClose={handleHideCreatePlayerModal}
                          onSubmit={joinGame} gameId={currentGameId}/>
            <Divider/>
            <div className="banner">
                <h1 className="left-title" style={{marginLeft: "10px"}}>Public games</h1>
                <Button variant="info" style={{marginRight: "10px", height: "40px", width: "100px", textAlign: "left"}}
                        onClick={handleClickRefresh} data-test="refresh-btn">
                    Refresh{" "}
                    {showRefreshSpinner && <Spinner
                        as="span"
                        animation="border"
                        size="sm"
                        role="status"
                        aria-hidden="true"
                        style={{marginBottom: "2px"}}
                    />}</Button>
            </div>
            <div className="container" data-test="lobby-cards-container">
                <div className="row">
                    {games.map(renderCard)}
                </div>
            </div>
        </div>
    )
}

export default Games;