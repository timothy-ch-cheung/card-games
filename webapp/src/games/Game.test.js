import React from 'react';
import {configure, mount} from 'enzyme';
import Games from "./Games";
import Adapter from 'enzyme-adapter-react-16';
import configureStore from "redux-mock-store";
import {Provider} from "react-redux";

configure({adapter: new Adapter()})

const mockStore = configureStore([]);
let store = mockStore({
    gameMode: "Match Two"
});

test('Matches Games snapshot', () => {
    const tree = mount(<Provider store={store}><Games/></Provider>);
    expect(tree).toMatchSnapshot();
});