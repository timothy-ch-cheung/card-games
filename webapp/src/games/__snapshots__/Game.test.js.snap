// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches Games snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Games>
    <div>
      <style
        type="text/css"
      >
        
                        .left-title {
                            text-align: left;
                        }
                        
                        .btn-xl {
                            font-size: 1.5rem;
                            border: grey solid 1px;
                            width: 200px;
                            margin: 20px 10px 30px;
                        }
                        
                        .container {
                            max-width: 98%;
                            margin: 0 auto;
                        }
                        
                        .row {
                            display: flex;
                            flex-flow: row wrap;
                        }

                    
      </style>
      <Button
        active={false}
        disabled={false}
        onClick={[Function]}
        size="xl"
        type="button"
        variant="info"
      >
        <button
          className="btn btn-info btn-xl"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Create Game
        </button>
      </Button>
      <CreateGame
        onClose={[Function]}
        show={false}
      />
      <CreatePlayer
        gameId={null}
        onClose={[Function]}
        onSubmit={[Function]}
        show={false}
      >
        <div>
          <style
            type="text/css"
          >
            
                        .modal-header button {
                           display: none;
                        }
                        
                        .btn-xxl {
                            font-size: 1.6rem;
                            border: grey solid 1px;
                            width: 220px;
                            margin: 20px 10px 30px;
                        }
                        
                        .left-title {
                            text-align: left;
                        }
                        
                        @media screen and (max-width: 768px) {
                            .btn-xxl {
                                font-size: 1.3rem;
                                width: 170px;
                            }
                        }
                        
                        .create-player-modal {
                            margin-top: 50px;
                        }
                        
                        input[type=text] {
                            margin-bottom: 5px;
                        }
                    
          </style>
          <Bootstrap(Modal)
            className="create-player-modal"
            onHide={[Function]}
            show={false}
          >
            <Modal
              animation={true}
              autoFocus={true}
              backdrop={true}
              bsPrefix="modal"
              className="create-player-modal"
              dialogAs={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "displayName": "ModalDialog",
                  "render": [Function],
                }
              }
              enforceFocus={true}
              keyboard={true}
              onHide={[Function]}
              restoreFocus={true}
              show={false}
            >
              <Modal
                autoFocus={true}
                backdrop={true}
                backdropTransition={[Function]}
                className="create-player-modal modal"
                containerClassName="modal-open"
                enforceFocus={true}
                keyboard={true}
                manager={
                  BootstrapModalManager {
                    "adjustAndStore": [Function],
                    "containers": Array [],
                    "data": Array [],
                    "handleContainerOverflow": true,
                    "hideSiblingNodes": true,
                    "modals": Array [],
                    "restore": [Function],
                    "scrollbarSize": 0,
                  }
                }
                onClick={[Function]}
                onEnter={[Function]}
                onEntering={[Function]}
                onExited={[Function]}
                onHide={[Function]}
                onMouseUp={[Function]}
                renderBackdrop={[Function]}
                restoreFocus={true}
                show={false}
                style={Object {}}
                transition={[Function]}
              />
            </Modal>
          </Bootstrap(Modal)>
        </div>
      </CreatePlayer>
      <styled.div>
        <div
          className="sc-AxjAm dcPIMr"
        />
      </styled.div>
      <h1
        className="left-title"
      >
        Public games
      </h1>
      <div
        className="container"
      >
        <div
          className="row"
        />
      </div>
    </div>
  </Games>
</Provider>
`;
